### PR TITLE
Converting temperature from Fahrenheit to Celsius 

### PR DESCRIPTION
**Pull request by chelseybaker**
Build on #7 

[Trello card](https://trello.com/c/8z8DL69e/34-convert-entire-app-to-display-celcius-instead-of-fahrenheit-do-not-make-changes-to-your-api-requests-to-achieve-this-instead-wri)

Converting temperatures from Fahrenheit to Celsius!

### CODE REVIEW

File: app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * Created by chelseybaker on 9/7/16.
+ */
+public class TemperatureConversionUtilityTest {
+
+    @Test
+    public void testFreezing() {
+        float celciusTemp = TemperatureConversionUtility.convertFahrenheit((float) 32.0);
+        assertEquals(celciusTemp, (float) 0, 0);
```
Comment (stkent) [inline app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java]: Shorthand for `(float) 0` is `0f`. Similarly for `100f`, `32.22f`.

Comment (stkent) [inline app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java]: Normally I think JUnit methods place the expected value before the computed value (check the method sig parameter names to verify this). If so, you'd want to swap your first two parameters for consistency.

File: app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    @Test
+    public void testBoiling() {
+        float celciusTemp = TemperatureConversionUtility.convertFahrenheit((float) 212);
+        assertEquals(celciusTemp, (float) 100, 0);
```
Comment (stkent) [inline app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java]: As above r.e. parameter ordering.

File: app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    @Test
+    public void testHotOut() {
+        float celciusTemp = TemperatureConversionUtility.convertFahrenheit((float) 90);
+        assertEquals(celciusTemp, (float) 32.22, 0.01);
```
Comment (stkent) [inline app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java]: As above r.e. parameter ordering.

File: app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package com.detroitlabs.chelseybaker.wundergroundweather.utilities;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Created by chelseybaker on 9/7/16.
+ */
+public class TemperatureConversionUtilityTest {
```
Comment (stkent) [inline app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java]: Good coverage! Love testing static utility methods :)

**Comment by chelseybaker** [comment]
Should be good now :)

File: app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java (hunk 1 above)
Comment (chelseybaker) [inline app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java]: :white

Comment (chelseybaker) [inline app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java]: ✅

File: app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java (hunk 3 above)
Comment (chelseybaker) [inline app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java]: ✅

File: app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java [hunk 5]
```diff
@@ -40,8 +49,37 @@ protected void onCreate(@Nullable final Bundle savedInstanceState) {
...
+                        .addApi(LocationServices.API)
+                        .build();
+           }
+        }
+    }
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        mGoogleApiClient.connect();
```
Comment (stkent) [inline app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java]: Oop, probably missed this in the last PR, but you can remove this call (and the `disconnect` in `onStop`) - the `enableAutoManage` method will make sure those get invoked during those lifecycle events automatically!

File: app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/WeatherForecastListViewActivity.java [hunk 6]
```diff
@@ -24,7 +24,13 @@
 
 public class WeatherForecastListViewActivity extends AppCompatActivity {
 
-    public static final String TEN_DAY_FORECAST = "com.detroitlabs.chelseybaker.wundergroundweather.TEN_DAY_FORECAST";
+    private static final String TEN_DAY_FORECAST = "com.detroitlabs.chelseybaker.wundergroundweather.TEN_DAY_FORECAST";
+
+    static public void start(Context context, TenDayForecast tenDayForecast) {
```
Comment (stkent) [inline app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/WeatherForecastListViewActivity.java]: Normal sequencing is `public static`

File: app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/models/ForecastDay.java [hunk 7]
```diff
@@ -50,12 +56,32 @@ public String getTitle() {
...
+        return convertTemperatures(forecastText);
     }
 
     public String getIcon_url() { return icon_url; }
 
     public ForecastDay() {
     }
 
+    private String convertTemperatures(String forecastText) {
+        Matcher m = Pattern.compile("\\d{1,3}F").matcher(forecastText);
```
Comment (stkent) [inline app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/models/ForecastDay.java]: If this was used more frequently, you'd want to compile once for performance reasons. See e.g. this decent discussion: http://stackoverflow.com/a/1721778/2911458

File: app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/models/ForecastDay.java [hunk 8]
```diff
@@ -50,12 +56,32 @@ public String getTitle() {
...
+            forecastText = forecastText.replaceAll(farenheight, celciusText);
+        }
+
+        m = Pattern.compile("\\d{1,3}s").matcher(forecastText);
+        while (m.find()) {
+            String farenheight = m.group();
+            Float celcius = TemperatureConversionUtility.convertFahrenheit(valueOf(farenheight.replace("s", "")));
+            String celciusText = String.format("%.0f", celcius).concat("s");
+            forecastText = forecastText.replaceAll(farenheight, celciusText);
+        }
```
Comment (stkent) [inline app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/models/ForecastDay.java]: What does this part of the method do?

File: app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java (hunk 4 above)
Comment (stkent) [inline app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java]: Per earlier comments, maybe try writing some tests with a negative Fahrenheit value in here...

File: app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/models/ForecastDay.java (hunk 8 above)
Comment (chelseybaker) [inline app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/models/ForecastDay.java]: The forecast text of the response would have the temperatures in fahrenheit, and the card said to convert all instances of fahrenheit to celsius, so it parses the string for digits followed by "s" (as in, "Tonight will be in the high *60s*") and converts that as well

File: app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java [hunk 9]
```diff
@@ -40,8 +49,37 @@ protected void onCreate(@Nullable final Bundle savedInstanceState) {
...
+
+        final Button searchByLocationButton = (Button)findViewById(R.id.search_current_location_button);
+        if (searchByLocationButton != null) {
+            searchByLocationButton.setOnClickListener(this);
+
+            // Create an instance of GoogleAPIClient.
+            if (mGoogleApiClient == null) {
+                mGoogleApiClient = new GoogleApiClient.Builder(this)
+                        .addConnectionCallbacks(this)
+                        .enableAutoManage(this, null) // Removes the need for GoogleApiClient.OnConnectionFailedListener
```
Comment (stkent) [inline app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java]: Now that we're using this, the client automatically connects in onStart. Before, we started that connection procedure when the button was clicked. As a result, the app is auto-advancing when the client connects. You'll want to remove the call to `advanceWithLastLocation` in `onConnected` and instead move it directly into the button click method.

File: app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java (hunk 4 above)
Comment (chelseybaker) [inline app/src/test/java/com/detroitlabs/chelseybaker/wundergroundweather/utilities/TemperatureConversionUtilityTest.java]: Done!

File: app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java (hunk 5 above)
Comment (chelseybaker) [inline app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java]: Makes sense haha

File: app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java (hunk 9 above)
Comment (chelseybaker) [inline app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java]: :nice:

Comment (stkent) [inline app/src/main/java/com/detroitlabs/chelseybaker/wundergroundweather/activities/LandingActivity.java]: Yep, this is a much more typical pattern - fire up the client connection based on lifecycle methods, then hook into it when UI interaction occurs :)

**Pull request merged**